Mesh updates and documentation

Thoemi09 opened the pull request:
Add documentation to meshes and do some clean up. The clean up is mostly trivial. If not, I have tried to make individual commits so that it easier to follow the changes.

Thoemi09 committed:
Add docs to mesh/tail_fitter.hpp and do minor clean up

- fix includes
- rename variables for docs and consistency

Thoemi09 committed:
Clean internal logic in mesh/tail_fitter.hpp to be consistent with the docs

- rename local variables
- use std::complex<double> instead of dcomplex
- do not rely on typedefs in other headers/namespaces
- simplify logic and add internal docs to make it easier to understand

Thoemi09 committed:
Reorganize class layouts in mesh/tail_fitter.hpp

- change tail_fitter_handle from a struct to a class
- no other code changes

Thoemi09 committed:
Rename class member variables in mesh::tail_fitter and mesh::tail_fitter_handle

Thoemi09 committed:
Symmetrize index ranges for tail fitting

- always include the lowest and highest frequencies in the tail fitting procedure

Thoemi09 committed:
Add docs to lattice/bravais_lattice.hpp and do minor clean up

- fix includes
- rename variables for docs and consistency
- add nodiscard and missing const

Thoemi09 committed:
Reorganize class layout of lattice::bravais_lattice

- change bravais_lattice::point_t from a struct to a class
- no other code changes

Thoemi09 committed:
Rename class member variables in lattice::bravais_lattice and lattice::bravais_lattice::point_t

Thoemi09 committed:
Minor improvements in lattice::bravais_lattice

- in operator==, compare orbital positions instead of number of orbitals
- improve output to std::ostream
- simplify logic in constructor

Thoemi09 committed:
Add docs to lattice/brillouin_zone.hpp and do minor clean up

- fix includes
- rename variables for docs and consistency
- add nodiscard and missing const
- remove copy/move operations and simplify constructor logic
- improve output to std::ostream
- fix bug in lattice_to_real_coordinates and real_to_lattice_coordicnates

Thoemi09 committed:
Rename class member variables in lattice::brillouin_zone

Thoemi09 committed:
Add docs to mesh/cyclat.hpp and do minor clean up

- fix includes
- rename variables for docs and consistency
- make evaluate functions non-friends

Thoemi09 committed:
Reorganize class layout of mesh::cyclat

- change cyclat::mesh_point_t from a struct to a class
- no other code changes

Thoemi09 committed:
Rename class member variables in mesh::cyclat and mesh::cyclat::mesh_point_t

- simplify logic in to_index(data_index_t)
- make to_data_index a non-template function

Thoemi09 committed:
Add docs to mesh/k_expr.hpp and do minor clean up

- fix includes
- add [[nodiscard]]

Thoemi09 committed:
Add docs to mesh/brzone.hpp and do minor clean up

- fix includes
- rename variables for docs and consistency

Thoemi09 committed:
Reorganize class layout of mesh::brzone

- change brzone::mesh_point_t from a struct to a class
- no other code changes

Thoemi09 committed:
Rename class member variables in mesh::brzone and mesh::brzone::mesh_point_t

- simplify logic in to_index(data_index_t)

Thoemi09 committed:
Make evaluate non-friend functions of mesh::brzone

- add a brzone::units_inv() member function

Thoemi09 committed:
Add docs to mesh/evaluate.hpp and do minor clean up

- fix includes
- rename variables for docs and consistency

Thoemi09 committed:
Adjust error tolerance in python/base/tail_issues.py

Wentzell reviewed (commented):
Thank you @Thoemi09 for this massive documentation improvement! I have merged a large number of the proposed changes into unstable, and will continue reviewing the rest in the next days.

Wentzell commented:
As discussed 5d62b592 is problematic due to floating point arithmetic, will probably skip these changes during merge

Thoemi09 committed:
Clean up utility/legendre, add docs and tests

Thoemi09 committed:
Add tests for mesh utilities

Thoemi09 committed:
Add tests for discrete mesh

Thoemi09 committed:
Minor updates to mesh::detail::linear

- be more restrictive in the constructor
- compiler generated operator!=
- remove unnecessary checks
- protect against N_ == 1

Thoemi09 committed:
Add tests for retime, imtime and refreq meshes

Thoemi09 committed:
Add tests for Matsubara frequencies

Thoemi09 committed:
Add tests for imfreq meshes

Thoemi09 committed:
Add tests for legendre meshes

Thoemi09 committed:
Add test for tail fitter

Thoemi09 committed:
Add tests for DLR meshes

Thoemi09 committed:
Add tests for Bravais lattice and Brillouin zone

Thoemi09 committed:
Add tests cyclat meshes

Thoemi09 committed:
Add tests brzone meshes

Thoemi09 committed:
Add tests for adjoint meshes

Thoemi09 committed:
Add tests for product meshes

Thoemi09 committed:
Remove old mesh tests

Thoemi09 committed:
Add tests for evaluating functions on meshes

Thoemi09 committed:
Restructure mesh docs

- add group Mesh types

Wentzell committed:
Fix bravais_lattice.cpp nda::linalg calls

Wentzell reviewed (approved):
Thank you @Thoemi09 for this massive improvement! This was now merged into unstable after rebase.
